### ISSUE TITLE
Support Nu Shell

### ISSUE THREAD

**Issue by andrewbanchich** [ajeetdsouza/zoxide#69]
I'd love to be able to use `zoxide` from [Nu Shell](https://github.com/nushell/nushell).

**Comment by cole-h** [comment]
I think the sentiment from the Ion shell issue can be echoed here (https://github.com/ajeetdsouza/zoxide/issues/66#issuecomment-613246939): a PR would be welcome, but I'm not familiar with Nu's syntax, and I don't think @ajeetdsouza is, either.

**Comment by ajeetdsouza** [comment]
Nushell lacks certain features that are necessary for zoxide to function, so it isn't possible at the moment to implement support for Nushell. I've created an issue on their tracker regarding the same.

The only workaround I've found will allow zoxide to work for users that have enabled the Starship prompt on Nushell. As it stands, I think implementing support for Nushell would only make sense if:
- Nushell adds support for PROMPT hooks
- Nushell removes the default prompt and makes Starship the only available prompt

As there isn't much else we can do from our end, I'm putting this issue on hold for now.

**Comment by mborejdo** [comment]
It seems there is progress on nushells prompt-configuration (https://github.com/nushell/nushell/pull/2064) . Is there anything holding this still back?

**Comment by ajeetdsouza** [comment]
Hey @mborejdo, we're still waiting on a couple more features from Nushell:

- If conditions (https://github.com/nushell/nushell/pull/2106), which should be released in v0.17.0.
- Command substitution (https://github.com/nushell/nushell/issues/2174). This is apparently a bug, and hopefully will be fixed soon.
- Variables (https://github.com/nushell/nushell/issues/291). This is part of an ongoing effort to make Nushell more scriptable, and my guess is that it would take a while. Currently, variables are used by zoxide to only cd upon a successful query, which looks like this:
  ```
  result="$(zoxide query args)" && cd "$result"
  ```
  One workaround might be to use `cd $(zoxide query args || pwd)`, which should work for most use cases. I can test this once the above bug is fixed.
- Variadic arguments for aliases (https://github.com/nushell/nushell/issues/1716). I suspect that we might be able to work around this by defining an alias with an absurdly large number of arguments, which no user is likely to exceed.
- Logical paths via `pwd` (https://github.com/nushell/nushell/issues/2175). This is not really a blocker; even now, we simply fallback to physical paths when the logical path is unavailable.
- Some kind of `eval` to run the `init` script. However, since Nushell only requires to be configured once, after which it works on all instances, we might be able to invoke `nu -c` from Rust itself to configure the hooks and aliases. In the worst case, we can ask the user to manually add the hook and aliases to their prompt.

I can't guarantee that this will be possible even after all of these are implemented, however, this is what I found so far.

**Comment by mborejdo** [comment]
hey @ajeetdsouza . thank you for the detailed explanation and tips for workarounds to make zoxide kindof workable for me. looking forward to a smoother installation and maybe hopefully your suggestion of some deeper integration within nushell itself.

**Comment by lpil** [comment]
Hello! This example config was shared [on twitter](https://twitter.com/mediacoder/status/1293627080148647948), it shows how zoxide can be used with Nu.

https://gist.github.com/mborejdo/65ddf0622dff641f2dab9e54167c11ee

I think the important bits are this:

```toml
startup = [
	"alias z [a b c d] {cd $(zoxide query $a $b $c $d | trim)}",
	"alias zi [a b c d] {cd $(zoxide query -i $a $b $c $d | trim)}",
	"alias za [a b c d] {zoxide add $a $b $c $d}",
	"alias zq [a b c d] {zoxide query $a $b $c $d}",
	"alias zqi [a b c d] {zoxide query -i $a $b $c $d}",
	"alias zr [a b c d] {zoxide remove $a $b $c $d}",
	"alias zri [a b c d] {zoxide remove -i $a $b $c $d}",
	"alias zoxship [a b c d] { do -i { zoxide add; starship prompt | echo `{{$(ansi white)}}{{$a}} {{$b}}{{$(ansi reset)}} {{$it}}` }}"
]
prompt = "zoxship "
```

**Comment by mborejdo** [comment]
hehe, thats part of my config, which makes zoxide "usable for my daily routine" (without knowing autojump or z very good).

The imporant bits are:
- the install process is manual
- currently there is a bug in nushell, which prevents evaluating output of external commands. using nushells trim command, works around this problem
- I did not want to run zoxide add as a starship-command, but still use starship. that's the reason for the strange "zoxide" alias.
- the broken character is some tux icon, as I use very similar shell-setups on windows and linux, and needed some visual feedback

**Comment by lpil** [comment]
Oh yes! Thank you once again :)

**Comment by ajeetdsouza** [comment]
Hey @mborejdo, thanks for finding a workaround! I'm sure this will be very helpful for anyone who wants to run zoxide on Nushell till it can be properly supported.

@lpil, I'm glad to know it's working for you too! I am also quite excited about Nushell, and I would love to see zoxide running on it. Without the features I mentioned above, however, such an implementation would be incomplete / inconsistent with the other shells, so I've decided to put this on hold till they are implemented.

**Comment by fdncred** [comment]
Here's a couple more aliases that I've been playing around with. Not perfect but a start!
```
    # does directory exist in zq? return true or false
    "alias dir_exists_in_z [a] {echo [$(if $(= $(do -i {zq $a}) | empty?) == $true {= $false} {= $true})]}",
    # if directory does not exists in zq, then use za to add it, and then z to cd to it, otherwise just use z
    "alias cdz [a] { if $(dir_exists_in_z $a) == $false {za $a ; z $a} { z $a } }",
    # tweak to support interactive mode via do command
    "alias zi [a b c d] {echo $(do -i {zoxide query -i $a $b $c $d} | trim) | cd $it}",
    # same tweak but without trim, not sure what diff it makes, haven't played enough
    "alias zqi [a b c d] {echo $(do -i {zoxide query -i $a $b $c $d}) | cd $it}",
    "alias za [a b c d] {zoxide add $a $b $c $d}",
    "alias z [a b c d] {cd $(zoxide query $a $b $c $d | trim)}",
```
I used @lpil 's gist as a starting point. I also use `cdz` instead of `z` right now.

**Comment by ajeetdsouza** [comment]
Hey @fdncred, thanks for the script! That's one more step towards official support :smile:

One minor correction though - `zoxide add` must be called every time you `cd` into a directory, not just when it's absent from the database! This is how zoxide scores each entry based on frequency of use and how recently you last used it. 

zoxide actually has two hooks for adding entries into the database:
- `prompt`: this calls `zoxide add` upon every prompt (i.e after every command is run). Nushell already supports prompt functions.
- `pwd` (default): this calls `zoxide add` every time the shell's working directory is changed. Calling `zoxide add` from the `z` alias is a good approximation of this, the only difference being that if the user uses a regular `cd`, that will not be registered in the database, which is not ideal behaviour. Nushell currently does not have a PWD hook that we can use, so we'll have to find a clever way of working around this.

**Comment by fdncred** [comment]
> One minor correction though - `zoxide add` must be called every time you `cd` into a directory, not just when it's absent from the database! This is how zoxide scores each entry based on frequency of use and how recently you last used it.

Oh ya, that makes sense. That should make it a bit easier I think. I'm update it and repost.

> * `prompt`: this calls `zoxide add` upon every prompt (i.e after every command is run). Nushell already supports prompt functions.

So, add `zoxide add $(pwd)` in the prompt configuration? or is there a special syntax for the prompt? I'll play with this tomorrow sometime.

> * Nushell currently does not have a PWD hook that we can use, so we'll have to find a clever way of working around this.

I'll have to think on this one a bit.

**Comment by fdncred** [comment]
Loving it! Here's my `prompt hook` with no changes to nushell.
`prompt = "za $(pwd) ; echo [ $(ansi gb) $(pwd) $(ansi reset) '(' $(ansi cb) $(do -i { git rev-parse --abbrev-ref HEAD  } | trim | str collect ) $(ansi reset) ')' $(char newline) $(ansi yb) $(date format '%m/%d/%Y %I:%M:%S%.3f %p' --raw) $(ansi reset) '> ' ] | str collect"`

Here's my complete list of zoxide aliases but `remove i` isn't working. Not sure exactly what it's supposed to do tbh.
```
    "alias z [a b c d] {cd $(zoxide query $a $b $c $d | trim)}",
    "alias zi [a b c d] {echo $(do -i {zoxide query -i $a $b $c $d} | trim) | cd $it}",
    "alias za [a b c d] {zoxide add $a $b $c $d}",
    "alias zq [a b c d] {zoxide query $a $b $c $d}",
    "alias zl [] {zoxide query --list}",
    "alias zqi [a b c d] {echo $(do -i {zoxide query -i $a $b $c $d}) | cd $it}",
    "alias zr [a b c d] {zoxide remove $a $b $c $d}",
    "alias zri [a b c d] {zoxide remove -i $a $b $c $d}",
```
![image](https://user-images.githubusercontent.com/343840/90984663-09985c80-e53c-11ea-9873-3f7c6eea52f6.png)

**Comment by ajeetdsouza** [comment]
Hey @fdncred,

I really appreciate your work on this! There are a couple of things left:

- As I mentioned in https://github.com/nushell/nushell/issues/2343, `z` (no arguments) must be equivalent to `cd ~`, and `z -` must be equivalent to `cd -` (jump to previous directory).
- Somehow, on my system, none of the interactive queries (`zi`, `zqi`, `zri`) are working.
- Regarding `zri` - `zoxide remove` requires a complete path to remove an entry from the shell. Since users rarely want to type in a complete path, it's a lot easier to have an interactive menu when removing a pesky entry from the DB. This will be a little more difficult to implement, since it is implemented entirely as a shell function using `zoxide query -i` under the hood. `zoxide remove -i` is not a valid command.

**Comment by fdncred** [comment]
> I really appreciate your work on this! 

You're welcome. This is fun!

> * As I mentioned in [nushell/nushell#2343](https://github.com/nushell/nushell/issues/2343), `z` (no arguments) must be equivalent to `cd ~`, and `z -` must be equivalent to `cd -` (jump to previous directory).

I think I can change the `z` alias to work this way.

> * Somehow, on my system, none of the interactive queries (`zi`, `zqi`, `zri`) are working.

Are you on linux/mac? I've only tested on windows.

> * `zoxide remove -i` is not a valid command.

ok. I  guess I was/am confused with the changelog entry saying, `Interactive mode for removing entries (zoxide remove -i).` I did see how it was implemented in powershell and I think I understand what needs to change, just query and then pass the results to remove.

**Comment by ajeetdsouza** [comment]
> Are you on linux/mac?

I tested on both Linux and Mac, and the same issue occurred. The terminal remains blank, and fzf doesn't show.

> I guess I was/am confused with the changelog entry

`zoxide remove -i` was removed in v0.4.2. The changelog entry wasn't clear enough, I've fixed it now!

**Comment by fdncred** [comment]
> I think I can change the `z` alias to work this way.

How's this:
`alias z [a b c d] {if $(echo $a | empty?) == $true { cd ~ } {echo $a | if $it == '-' { cd - } {cd $(zoxide query -- $a $b $c $d | trim)}}}`

**Comment by ajeetdsouza** [comment]
That's incredible! Two more changes would make it fully compliant with other shells:

- `z - foo` will still go to the previous directory, when it should actually be executing `zoxide query -- - foo`.
- When the environment variable `_ZO_ECHO` is set, `z` echoes the new `$PWD` value. If I understand correctly, Nushell doesn't exactly use environment variables for configuration, rather, it seems to rely on `config set`. So accordingly, if the value of `config get _ZO_ECHO` is `1`, we should echo `$PWD` after changing the directory.

**Comment by fdncred** [comment]
> * `z - foo` will still go to the previous directory, when it should actually be executing `zoxide query -- - foo`.

ok, so for this one, I'd have to check the second parameter to see if it should `cd -` or `zq - foo`

> _ZO_ECHO

I'm not sure what to do about that. I can check for it like this `if $nu.env._ZO_ECHO == "1" { echo $true } { echo $false }` but if it doesn't exist at all it pukes.

**Comment by ajeetdsouza** [comment]
I came up with a _very_ hacky way to do initialization, which should take care of this. It shouldn't have too much impact on performance, but the advantage of something this dynamic is that whenever zoxide updates, the user automatically gets updated aliases in their shell.

To initialize zoxide on nushell, the user simply calls `zoxide init nu` on any shell. From Rust, we do this:

- Run `nu -c 'echo $nu.env.HOME'`. If this fails with exit code `1`, that means `_ZO_ECHO` was not set. Run a `nu -c` command again to set it permanently (I couldn't find the exact command to do this).
- Run `nu -c 'config get startup | str find-replace "^alias (z|zi) .*$" "" | where $it not-in: ["" "zoxide init nu"] | append "zoxide init nu" | append "alias z [msg] { echo $msg }" | append "alias zi [msg] { echo $msg }" | config set_into startup'`. I've set dummy aliases here, but you get the idea - remove all previous aliases set by zoxide from `startup`, and add new ones. Also add `zoxide init nu` to `startup` so that the configuration is regenerated whenever a new session is created.

**Comment by ajeetdsouza** [comment]
I ran `hyperfine` on `nu -c ""` and the performance hit is around 25ms. Since our startup script calls `nu` thrice, we're looking at a > 75ms impact on shell loading times using this method.

We could just leave out `zoxide init nu` from the startup variable, but that means the user has to remember to call `zoxide init nu` every time zoxide is updated, or things could break.

**Comment by fdncred** [comment]
Wow! Cool! Yes, hackish but if it works, it works. For users in general I think this will work. My only minor reservation in going this route, is I will never run `zoxide init nu` because nushell the comments in my toml file would get overwritten. I've been working on using the `toml_edit` crate which works as expected and maintains comments, but i haven't finished it yet. Other than that, I think this is a cool solution.

btw - i did some minor testing in wsl2 yesterday and saw the zi/zqi from within nu and i'm not sure why it's not showing fzf. it works in wsl2 in zsh/bash/whatever. so, i'm puzzled.

**Comment by ajeetdsouza** [comment]
> the comments in my toml file would get overwritten

Hmm, never thought about that. Yes, `toml_edit` definitely sounds like a good idea for Nushell.

---

I missed this earlier, but it seems Nushell does provide a simpler way to create aliases:

```
alias z --save [msg] { echo $msg }
```

This could be a bug, but I noticed if I create an alias `z` using `config set startup`, and then create another one using `alias z --save`, it does not overwrite the previous alias, but if I call `alias z --save` a second time, it overwrites its own previously set alias.

**Comment by fdncred** [comment]
> I missed this earlier, but it seems Nushell does provide a simpler way to create aliases:

LOL! You'd think that I'd remember this however, I always edit my aliases with vscode. I honestly can't remember all the features we add.

> a second time, it overwrites its own previously set alias.

Can you file that as a bug. Sounds like a bug to me.

**Comment by ajeetdsouza** [comment]
> Sounds like a bug to me.

Created an issue (https://github.com/nushell/nushell/issues/2419).

---

It seems interactive queries are blocked due to a bug too, but that is a non-critical feature. The only major thing remaining is figuring out hooks. If we do that, I think we can add beta support for Nushell to zoxide.

**Comment by fdncred** [comment]
I have prompt hooks working via item in config startup.

**Comment by ajeetdsouza** [comment]
Could you share the script? My two main concerns are:

- Does it work even if the user is using starship?
- Can the hook be created automatically?

**Comment by fdncred** [comment]
It's up above in my prompt config setting.

**Comment by ajeetdsouza** [comment]
Ah, I forgot about that. I had to make a small change to get it to work, though - `date format` needed to be changed to `date --format`. Other than that, it seems fine. To answer my own questions:

> Does it work even if the user is using starship?

Nope.

> Can the hook be created automatically?

If Starship isn't configured, then yes.

---

Here's what we could do:

- If `nu -c 'config get use_starship'` is `true`, then we ask the user to add `zoxide add $(pwd)` as a [Starship custom command](https://starship.rs/config/#custom-commands) to his config file.
- Otherwise, we search for `zoxide add $(pwd)` in `config get prompt`, and if it isn't there, we add it in.

Getting this to play well with the rest of the `init` module is going to require quite a few changes. @fdncred, let me know if you want to create a draft PR with a POC - I'd be happy to help out!

**Comment by mborejdo** [comment]
Hey all. Great to see some progress here. @ajeetdsouza I am wondering, if `use_starship` will be deprecated, as nushell seems to be removing starship integration in favor for the prompt-hook which can run starship binary itself (see https://github.com/nushell/nushell/pull/2345)

**Comment by ajeetdsouza** [comment]
If starship is being phased out, we don't need to support it. That's great news, it means zoxide can get fully automated initialization for prompt hooks.

**Comment by fdncred** [comment]
@mborejdo you are correct. @ajeetdsouza `use_starship` doesn't exist anymore. we no longer bundle starship with nushell. If a user wants to use starship with nushell they have to download and install starship.

**Comment by fdncred** [comment]
I just posted this in Nushell Discord #help channel but thought I'd put it here too if anyone needs help with configuring `zoxide` for nushell `0.24.2` (and probably `0.25`.

Add these nushell custom commands to the nushell `config.toml` in the `[startup]` section. These are kind of the bare-bones `zoxide` things to get it working. I'm on Windows using `zoxide v0.4.3-unknown`

```
    "def z [a] {if $(echo $a | empty?) == $true { cd ~ } {echo $a | each { if $it == '-' { cd - } {cd $(zoxide query -- $it | str trim)}}}}",
    "def zi [a] {echo $(do -i {zoxide query -i $a} | str trim) | each { cd $it }}",
    "def za [a:block] {zoxide add $a}",
    "def zq [a] {zoxide query -- $a}",
    "def zl [] {zoxide query --list}",
    "def zqi [a] {echo $(do -i {zoxide query -i $a | str trim }) | each { cd $it }}",
    "def zr [a] {zoxide remove $a}",
    "def zri [a] {zoxide remove -i $a}",
```
but the real key is to run `zoxide` in the prompt so it runs every time you change dir. my prompt looks like this, granted, you probably won't like my prompt. 👍 
```
    prompt = "za $(pwd) | = `{{$(set_title)}} {{$(git_br)}}`"
```

and these last two nushell custom commands are for `set_title` and `git_br` in the prompt. they go in the `[startup]` section too.

```
    "def set_title_str [str-arg] {echo [$(char title) ' ' $str-arg ' ' $(char bel)] | str collect}",
    "def git_br [] {echo [ $(ansi gb) $(pwd) $(ansi reset) '(' $(ansi cb) $(do -i { git rev-parse --abbrev-ref HEAD  } | str trim | str collect ) $(ansi reset) ')' $(char newline) $(ansi yb) $(date now | date format '%m/%d/%Y %I:%M:%S%.3f %p') $(ansi reset) ' ¯\\_(?)_/¯ ' $(char prompt) ' '] | str collect}",
```

**Comment by ajeetdsouza** [comment]
@fdncred there's one check missing from `z`. I don't know how easy it is to add, though.

If `z` has only one argument AND that argument is a cd-able path, `cd` there. Otherwise, call `zoxide query`. Test cases:
- `z .` should do nothing
- `z ..` should go to the parent directory
- `z /some/path/not/in/db` should work even though the path isn't in the database

The idea is to completely replace `cd` by `z`, you shouldn't have to call `cd` once zoxide is initialized.

**Comment by fdncred** [comment]
@ajeetdsouza Thanks. This change should be possible in the future since the `0.25.0` release will have proper scripting support. However, we have one fiddly thing that still needs to be fixed. Right now parameters are required. So, there's no option of calling `z` without parameters if it's defined as `z [param] {}`. I hope we'll get that fixed soon.

**Comment by ahkrr** [comment]
@fdncred to get around the limitation of not being able to call `nu` functions without arguments, in nushell 26 we can use a little hack to make this work 

    -> write an implementation-function
    -> make an alias always call the implementation-function with at least an empty string 
    -> make the function implementation skip the first empty string and assign the rest to a variable
    -> then profit

``` 
def z_impl [...rest:string] {let a = $(echo $rest | skip 1); if $(echo $a | empty?) == $true { cd ~ } {echo $a | each { if $it == '-' { cd - } {cd $(zoxide query -- $it | str trim)}}}}
alias z = z_impl ""
```

**Comment by ahkrr** [comment]
@fdncred I have rewritten the functions that you posted with the no arguments hack and the changes suggested by @ajeetdsouza  in mind.

In config.toml
```
startup = [
    "def __zoxide_z [...rest:string] { let input = $(echo $rest | skip 1); if $(echo $input | first 1 | empty?) == $true { cd ~ } { let first = $( echo $input | first 1 | str trim ); if  $( echo $input | count ) == 1 { if $first == '.' { } { if $first == '..' { cd .. } { if $first == '-' { cd - } { if $( echo $first | path exists ) { cd $first } { cd $( zoxide query $input | str trim ) } }}} } { cd $( zoxide query $input | str trim ) }}} ",
    "def __zoxide_zi [...rest:string] { let input = $(echo $rest | skip 1); cd $( zoxide query -i -- $input | str trim ); } ",
    "def __zoxide_za [...rest:string] { let input = $(echo $rest | skip 1); zoxide add $input; } ",
    "def __zoxide_zq [...rest:string] { let input = $(echo $rest | skip 1); zoxide query $input; } ",
    "def __zoxide_zqi [...rest:string] { let input = $(echo $rest | skip 1); zoxide query -i $input; } ",
    "def __zoxide_zr [...rest:string] { let input = $(echo $rest | skip 1); zoxide remove $input; } ",
    "def __zoxide_zri [...rest:string] { let input = $(echo $rest | skip 1); zoxide remove -i $input; } ",
    "alias z   = __zoxide_z   '' ",
    "alias zi  = __zoxide_zi  '' ",
    "alias za  = __zoxide_za  '' ",
    "alias zq  = __zoxide_zq  '' ",
    "alias zqi = __zoxide_zqi '' ",
    "alias zr  = __zoxide_zr  '' ",
    "alias zri = __zoxide_zri '' ",
    "def set_title_str [str-arg] {echo [$(char title) ' ' $str-arg ' ' $(char bel)] | str collect}",
    "def git_br [] {echo [ $(ansi gb) $(pwd) $(ansi reset) '(' $(ansi cb) $(do -i { git rev-parse --abbrev-ref HEAD  } | str trim | str collect ) $(ansi reset) ')' $(char newline) $(ansi yb) $(date now | date format '%m/%d/%Y %I:%M:%S%.3f %p') $(ansi reset) ' ¯\\_(?)_/¯ ' $(char prompt) ' '] | str collect}",
]
```
to copy and paste 
```
def __zoxide_z [...rest:string] { let input = $(echo $rest | skip 1); if $(echo $input | first 1 | empty?) == $true { cd ~ } { let first = $( echo $input | first 1 | str trim ); if  $( echo $input | count ) == 1 { if $first == '.' { } { if $first == '..' { cd .. } { if $first == '-' { cd - } { if $( echo $first | path exists ) { cd $first } { cd $( zoxide query $input | str trim ) } }}} } { cd $( zoxide query $input | str trim ) }}}
def __zoxide_zi [...rest:string] { let input = $(echo $rest | skip 1); cd $( zoxide query -i -- $input | str trim ); }
def __zoxide_za [...rest:string] { let input = $(echo $rest | skip 1); zoxide add $input; }
def __zoxide_zq [...rest:string] { let input = $(echo $rest | skip 1); zoxide query $input; }
def __zoxide_zqi [...rest:string] { let input = $(echo $rest | skip 1); zoxide query -i $input; }
def __zoxide_zr [...rest:string] { let input = $(echo $rest | skip 1); zoxide remove $input; }
def __zoxide_zri [...rest:string] { let input = $(echo $rest | skip 1); zoxide remove -i $input; }
alias z   = __zoxide_z   ''
alias zi  = __zoxide_zi  ''
alias za  = __zoxide_za  ''
alias zq  = __zoxide_zq  ''
alias zqi = __zoxide_zqi ''
alias zr  = __zoxide_zr  ''
alias zri = __zoxide_zri ''
```
I hope that this should now work like the other `zoxide` shell helpers
I don't know how I can manually  adjust the `prompt` in `nushell`. 
The `prompt variable` doesn't seem to work for me. Maybe I am doing something wrong.

**Comment by fdncred** [comment]
@ahkrr Nice! Very nice! I've been thinking about this. Another option is to use flags/switches which are optional. I'm not saying your work needs to change. I'm just throwing out another strategy. For instance, if you look at my `ls-wide.nu` script in the `nu_scripts` repo you'll see this:
```rust
# A ls command that approximates the ls -sh command in bash
def ls-wide [
    --columns(-c):int # The number of columns in your output
    ] {
    let is_empty = $(= $columns | empty?)
    let max_fname_size = $(ls | get name | str from | str length | math max)
    let max_fsize_size = $(ls | get size | str from | str length | math max)

    ls | each -n {
        build-string $(echo $it.item.name | str rpad -c ' ' -l $max_fname_size) ' ' $(echo $(build-string $it.item.size) | str lpad -c ' ' -l $max_fsize_size) ' '
        if $is_empty {
            if $(= $it.index + 1) mod 3 == 0 {
                echo $(char newline) | autoview
            } {}
        } {
            if $(= $it.index + 1) mod $columns == 0 {
                echo $(char newline) | autoview
            } {}
        }
    } | str collect
}
```
Note how `-c|--columns` is checked if it's provided or not.

Having said all this, I'm going to try out your work.  As a side note, here's some prompt statements you can try. I have like 20 of them.
```
prompt = "= `{{$(za $(pwd))}} {{$(set_title)}} {{$(git_br)}}`"
```

```
prompt = "za $(pwd) | starship prompt"
```

```
prompt = 'za $(pwd) | echo `{{$(set_title)}} {{$(git_br)}}`'
```

**Comment by fdncred** [comment]
I think we still may need to do this https://github.com/ajeetdsouza/zoxide/issues/69#issuecomment-679346038

**Comment by ahkrr** [comment]
Now the case with _ZO_ECHO == '1' is also handled.
The other issue of https://github.com/ajeetdsouza/zoxide/issues/69#issuecomment-679346038 should also not be a problem
config.toml 
```
startup = [
    "def __zoxide_cd [ ] { if $(let ZO_ECHO = $(do -i { echo $nu.env._ZO_ECHO }); echo $ZO_ECHO | empty? ) == $false { if $ZO_ECHO == '1' { echo $(pwd) } {}} {}}",
    "def __zoxide_z_impl [...rest:string] { let input = $(echo $rest | flatten | skip 1); if $(echo $input | first 1 | empty?) == $true { cd ~ } { let first = $( echo $input | first 1 | str trim ); if  $( echo $input | count ) == 1 { if $first == '.'  { } { if $first == '..' { cd .. } { if $first == '-'  { cd - } { if $( echo $first | path exists ) { cd $first } { cd $( zoxide query -- $input | str trim ) } }}} } { cd $( zoxide query -- $input | str trim ) } } } ",
    "def __zoxide_z   [...rest:string] { __zoxide_z_impl $rest; __zoxide_cd; } ",
    "def __zoxide_zi  [...rest:string] { let input = $(echo $rest | skip 1); cd $( zoxide query -i -- $input | str trim ); __zoxide_cd;} ",
    "def __zoxide_za  [...rest:string] { let input = $(echo $rest | skip 1); zoxide add       $input; } ",
    "def __zoxide_zq  [...rest:string] { let input = $(echo $rest | skip 1); zoxide query     $input; } ",
    "def __zoxide_zqi [...rest:string] { let input = $(echo $rest | skip 1); zoxide query -i  $input; } ",
    "def __zoxide_zr  [...rest:string] { let input = $(echo $rest | skip 1); zoxide remove    $input; } ",
    "def __zoxide_zri [...rest:string] { let input = $(echo $rest | skip 1); zoxide remove -i $input; } ",
    "alias z   = __zoxide_z   '' ",
    "alias zi  = __zoxide_zi  '' ",
    "alias za  = __zoxide_za  '' ",
    "alias zq  = __zoxide_zq  '' ",
    "alias zqi = __zoxide_zqi '' ",
    "alias zr  = __zoxide_zr  '' ",
    "alias zri = __zoxide_zri '' ",
]

prompt = "za; build-string $(ansi gb) $(pwd) $(ansi reset) '(' $(ansi cb) $(do -i { git rev-parse --abbrev-ref HEAD } | str trim ) $(ansi reset) ')' $(ansi yb) $(date format '%m/%d/%Y %I:%M:%S%.3f %p' ) $(ansi reset) '> ' "
```
to copy paste 
``` sh
def __zoxide_cd [ ] {
    if $(let ZO_ECHO = $(do -i { echo $nu.env._ZO_ECHO }); echo $ZO_ECHO | empty? ) == $false {
        if $ZO_ECHO == '1' { echo $(pwd) } { }
    } { }
}
def __zoxide_z [...rest:string] {
    __zoxide_z_impl $rest
    __zoxide_cd
}

def __zoxide_z_impl [...rest:string] {
    let input = $(echo $rest | flatten | skip 1);
    if $(echo $input | first 1 | empty?) == $true {
        cd ~
    } {
        let first = $( echo $input | first 1 | str trim );
        if  $( echo $input | count ) == 1 {
            if $first == '.'  { } {
            if $first == '..' { cd .. } {
            if $first == '-'  { cd - } {
            if $( echo $first | path exists ) {
                cd $first
            } {
                cd $( zoxide query -- $input | str trim )
            }
            }}}
        } {
            cd $( zoxide query -- $input | str trim )
        }
    }
}

def __zoxide_zi  [...rest:string] { let input = $(echo $rest | skip 1); cd $( zoxide query -i -- $input | str trim ); __zoxide_cd;}
def __zoxide_za  [...rest:string] { let input = $(echo $rest | skip 1); zoxide add       $input; }
def __zoxide_zq  [...rest:string] { let input = $(echo $rest | skip 1); zoxide query     $input; }
def __zoxide_zqi [...rest:string] { let input = $(echo $rest | skip 1); zoxide query -i  $input; }
def __zoxide_zr  [...rest:string] { let input = $(echo $rest | skip 1); zoxide remove    $input; }
def __zoxide_zri [...rest:string] { let input = $(echo $rest | skip 1); zoxide remove -i $input; }

alias z   = __zoxide_z   ''
alias zi  = __zoxide_zi  ''
alias za  = __zoxide_za  ''
alias zq  = __zoxide_zq  ''
alias zqi = __zoxide_zqi ''
alias zr  = __zoxide_zr  ''
alias zri = __zoxide_zri ''
```

**Comment by ajeetdsouza** [comment]
@ahkrr this is great work, thank you!

Small nit, though - you don't actually need to check for `if $first == '.'` - in fact, `z .` is not the same as doing nothing. Say, I'm currently in a directory, and run a script that deletes and recreates that directory. After this, commands like `ls` (anything that relies on getting the current directory) will fail. In situations like this, `z .` will get you back into the correct directory.

**Comment by ahkrr** [comment]
I understand the point, but not how to solve this. With a rough explanation on how to build it, I could figure out the implementation details.

**Comment by fdncred** [comment]
@ajeetdsouza are you saying just remove those checks like this?
```sh
def __zoxide_z_impl [...rest:string] {
    let input = $(echo $rest | flatten | skip 1);
    if $(echo $input | first 1 | empty?) == $true {
        cd ~
    } {
        let first = $( echo $input | first 1 | str trim );
        if  $( echo $input | count ) == 1 {
            if $first == '-'  { cd - } {
                if $( echo $first | path exists ) {
                    cd $first
                } {
                    cd $( zoxide query -- $input | str trim )
                }
            }
        } {
            cd $( zoxide query -- $input | str trim )
        }
    }
}
```

**Comment by ajeetdsouza** [comment]
Yup. `.` and `.. ` both return `True` when passed to `path exists`.

**Comment by fdncred** [comment]
cool!

ok, I'm thinking we have all of `zoxide` implemented for nushell then. If there's something missing please let us know. 

probably the best thing to do for an `init` routine is to write this stuff to a file and `source` it in nushell. That way, nothing in the `config.toml` gets overwritten.

**Comment by ajeetdsouza** [comment]
This is great, thanks to everyone who helped with this! :smile: I will work on adding this to zoxide today.

On a side note, since I'm not too familiar with Nushell syntax, do put a +1 on this post if it's okay for me to tag you in Nushell-related issues in future. I'd really appreciate the help!  @andrewbanchich @mborejdo @lpil @fdncred @ahkrr

**Comment by ahkrr** [comment]
@ajeetdsouza 
In `nushell` the name of the `count` `function` was changed to `length`. [#3166](https://github.com/nushell/nushell/pull/3166)
This will probably be included in the `nushell 0.29` release in a couple of days. 
This would have to be changed in the `scripts` that in this issue. 
In the documentation it should probably also be mentioned that these will work with `nushell  >= 0.29`

**Comment by ajeetdsouza** [comment]
@ahkrr thanks for the heads-up! I'll change this once `nushell 0.29` is out.

I've drafted a [PR](https://github.com/ajeetdsouza/zoxide/pull/164) for this, do check it out if you have a Rust compiler handy!

**Comment by ajeetdsouza** [comment]
Nushell support is finally merged. Thanks to everyone who helped with this!
